includes/class-newspack-listings-taxonomies.php

```diff
@@ -0,0 +1,451 @@
+				if ( empty( $term_slug ) || ! in_array( $term_slug, array_keys( self::NEWSPACK_LISTINGS_TAXONOMIES ) ) ) {
+					continue;
+				}
+
+				// Check for a shadow term associated with this post.
+				$shadow_term = self::get_shadow_term( $post, self::NEWSPACK_LISTINGS_TAXONOMIES[ $term_slug ] );
+
+				// If there isn't already a shadow term, create it. Otherwise, apply the term to the post.
+				if ( empty( $shadow_term ) ) {
+					self::create_shadow_term( $post, self::NEWSPACK_LISTINGS_TAXONOMIES[ $term_slug ] );
+				} else {
+					wp_set_post_terms( $post_id, $shadow_term->term_id, self::NEWSPACK_LISTINGS_TAXONOMIES[ $term_slug ] );
+				}
+			}
+		}
+	}
+
+	/**
+	 * Get related listing posts for the given shadow taxonomy terms by term slug.
+	 *
+	 * @param array $slugs Array of term slug to use for looking up post.
+	 * @return array Array of related listing posts.
+	 */
+	public static function get_related_listings( $slugs ) {
+		$related_listings = new \WP_Query(
+			[
+				'post_type'      => self::get_post_types_to_shadow(),
+				'posts_per_page' => 100,
+				'post_status'    => 'publish',
+				'post_name__in'  => $slugs,
+				'no_found_rows'  => true,
+				'fields'         => 'ids',
+				'order'          => 'ASC',
+				'orderby'        => 'type title',
+			]
+		);
+
+		if ( empty( $related_listings->posts ) || is_wp_error( $related_listings ) ) {
+			return [];
+		}
+
+		return $related_listings->posts;
+	}
+
+	/**
+	 * If the post has been assigned any listing shadow terms, append a link to the listing at the end of the content.
+	 *
+	 * @param string $content Post content.
+	 * @return string The filtered post content.
+	 */
+	public static function maybe_append_related_listings( $content ) {
+		$post_id       = get_the_ID();
+		$listing_terms = wp_get_post_terms( $post_id, array_values( self::NEWSPACK_LISTINGS_TAXONOMIES ) );
+
+		if ( 0 < count( $listing_terms ) ) {
+			$related_listing_ids = self::get_related_listings( array_column( $listing_terms, 'slug' ) );
+
+			if ( 0 < count( $related_listing_ids ) ) {
+				$related_section_title = Core::is_listing( get_post_type() ) ? __( 'Listed by', 'newspack-listings' ) : __( 'Related listings', 'newspack-listings' );
+				$content              .= '<hr class="wp-block-separator is-style-wide newspack-listings__separator" />';
+				$content              .= '<h3 class="accent-header newspack-listings__related-section-title">' . $related_section_title . '</h3>';
+			}
+
+			foreach ( $related_listing_ids as $related_listing_id ) {
+				$listing_title   = get_the_title( $related_listing_id );
+				$listing_excerpt = Utils\get_listing_excerpt( get_post( $related_listing_id ) );
+				$listing_url     = get_permalink( $related_listing_id );
+				$featured_image  = get_the_post_thumbnail( $related_listing_id, 'thumbnail', [ 'class' => 'avatar' ] );
+
+				$content .= '<div class="author-bio sponsor-bio newspack-listings__related-listing">'; // <a href="'. $listing_url . '">';
```

Comment:
The `.sponsor-bio` styles are actually set in the Newspack theme, and it makes the image a tiny bit larger than the regular `.author-bio` styles. It also removes the border radius to make the avatars appear circular (which shouldn't apply to these listings). I used these existing theme styles so that these elements would hopefully inherit styles from the child themes as well as theme customizations such as colors, but I'm open to rendering markup and styles specific to the plugin if you think that would be a better solution.